File: Singleton/Carrinho.js
const Carrinho = (function() {
    let instance;

    function createInstance() {
        let produtos = [];
        
        function addProduto(produto){
            produtos.push(produto);
        }

        function getProduto() {
            return produtos;
        }

        function clearCarrinho(){
            produtos = [];
        }

        return {
            addProduto, getProduto, clearCarrinho
        }
    }

    return {
        getInstance: () => {
            if(!instance){
                instance = createInstance();
            }
            return instance;
        }
    }
})()

const carrinho = Carrinho.getInstance();

carrinho.addProduto({nome: "Produto 1", preco: 10})
carrinho.addProduto({nome: "Produto 2", preco: 20})
carrinho.addProduto({nome: "Produto 3", preco: 30})
carrinho.addProduto({nome: "Produto 4", preco: 40})

console.log(carrinho.getProduto());

carrinho.clearCarrinho();

console.log(carrinho.getProduto());

carrinho.addProduto({nome: "Produto 5", preco: 15})
carrinho.addProduto({nome: "Produto 6", preco: 25})
carrinho.addProduto({nome: "Produto 7", preco: 35})
carrinho.addProduto({nome: "Produto 8", preco: 45})
console.log(carrinho.getProduto());

File: Singleton/Carrinho.py
class Carrinho:

    __instance = None

    def __init__(self):
        if Carrinho.__instance is None:
            Carrinho.__instance = self
            self.items = []
        else:
            raise Exception("Esta classe é um singleton")
    
    @staticmethod
    def get_instance():
        if not Carrinho.__instance:
            Carrinho()
        return Carrinho.__instance

    def add_item(self, item):
        self.items.append(item)

    def get_item(self):
        return self.items


class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco


if __name__ == '__main__':
    # Instanciando carrinho
    carrinho1 = Carrinho.get_instance()

    # Instanciando produtos
    produto1 = Produto("Banana", 4.00)
    produto2 = Produto("Maça", 7.99)
    produto3 = Produto("Melao", 3.89)
    produto4 = Produto("Repolho", 3.79)
    produto5 = Produto("Agua", 2.00)
    produto6 = Produto("Melancia", 1.99)

    # Adicionando no carrinho
    carrinho1.add_item(produto1)
    carrinho1.add_item(produto1)
    carrinho1.add_item(produto3)
    carrinho1.add_item(produto6)
    carrinho1.add_item(produto1)

    # Mostra na tela
    valorTotal = 0

    print("---------- Carrinho ----------")
    for item in carrinho1.get_item():
        print(f"{item.nome} - {item.preco}")
        valorTotal += item.preco

    print(f"{len(carrinho1.get_item())} items no carrinho")
    print(f"TOTAL: {round(valorTotal, 2)}")